srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+);
+
+impl<T: Trait> Module<T> {
+	// PUBLIC IMMUTABLES
+
+	/// Get an account's total balance of an asset kind.
+	pub fn total_balance(asset_id: &T::AssetId, who: &T::AccountId) -> T::Balance {
+		Self::free_balance(asset_id, who) + Self::reserved_balance(asset_id, who)
+	}
+
+	/// Get an account's free balance of an asset kind.
+	pub fn free_balance(asset_id: &T::AssetId, who: &T::AccountId) -> T::Balance {
+		<FreeBalance<T>>::get(asset_id, who)
+	}
+
+	/// Get an account's reserved balance of an asset kind.
+	pub fn reserved_balance(asset_id: &T::AssetId, who: &T::AccountId) -> T::Balance {
+		<ReservedBalance<T>>::get(asset_id, who)
+	}
+
+	/// Creates an asset.
+	///
+	/// # Arguments
+	/// * `asset_id`: An ID of a reserved asset.
+	/// If not provided, a user-generated asset will be created with the next available ID.
+	/// * `from_account`: The initiator account of this call
+	/// * `asset_options`: Asset creation options.
+	///
+	pub fn create_asset(
+		asset_id: Option<T::AssetId>,
+		from_account: Option<T::AccountId>,
+		options: AssetOptions<T::Balance, T::AccountId>,
+	) -> Result {
+		let asset_id = if let Some(asset_id) = asset_id {
+			ensure!(!<TotalIssuance<T>>::exists(&asset_id), "Asset id already taken.");
+			ensure!(asset_id < Self::next_asset_id(), "Asset id not available.");
+			asset_id
+		} else {
+			let asset_id = Self::next_asset_id();
+			let next_id = asset_id
+				.checked_add(&One::one())
+				.ok_or_else(|| "No new user asset id available.")?;
+			<NextAssetId<T>>::put(next_id);
+			asset_id
+		};
+
+		let account_id = from_account.unwrap_or_default();
+		let permissions: PermissionVersions<T::AccountId> = options.permissions.clone().into();
+
+		<TotalIssuance<T>>::insert(asset_id, &options.initial_issuance);
+		<FreeBalance<T>>::insert(&asset_id, &account_id, options.initial_issuance);
+		<Permissions<T>>::insert(&asset_id, permissions);
+
+		Self::deposit_event(RawEvent::Created(asset_id, account_id, options));
+
+		Ok(())
+	}
+
+	/// Transfer some liquid free balance from one account to another.
+	/// This will not emit the `Transferred` event.
+	pub fn make_transfer(asset_id: &T::AssetId, from: &T::AccountId, to: &T::AccountId, amount: T::Balance) -> Result {
+		let new_balance = Self::free_balance(asset_id, from)
+			.checked_sub(&amount)
+			.ok_or_else(|| "balance too low to send amount")?;
+		Self::ensure_can_withdraw(asset_id, from, amount, WithdrawReason::Transfer, new_balance)?;
+
+		if from != to {
+			<FreeBalance<T>>::mutate(asset_id, from, |balance| *balance -= amount);
+			<FreeBalance<T>>::mutate(asset_id, to, |balance| *balance += amount);
+		}
+
+		Ok(())
+	}
+
+	/// Transfer some liquid free balance from one account to another.
+	/// This will emit the `Transferred` event.
+	pub fn make_transfer_with_event(
+		asset_id: &T::AssetId,
+		from: &T::AccountId,
+		to: &T::AccountId,
+		amount: T::Balance,
+	) -> Result {
+		Self::make_transfer(asset_id, from, to, amount)?;
+
+		if from != to {
```

Comment:
why not move this up a couple of lines to it encompasses the entire function?

Reply:
the transaction should fail if the from account does not have enough balances, no matter if the receiver is the same or not.